File: EasyApp/EasyApp/Template/MVC2/MVC2Base.m
//
//  MVC2Config.m
//  EasyApp
//
//  Created by 靳小飞 on 2018/1/29.
//  Copyright © 2018年 靳小飞. All rights reserved.
//

#import "MVC2Base.h"

@interface BinderHelper : NSObject
+ (instancetype)shareBinderHelp;
@property (nonatomic, strong) NSMutableDictionary *binderMap;
@end

@implementation BinderHelper
+ (instancetype)shareBinderHelp
{
    static BinderHelper *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[BinderHelper alloc] init];
        instance.binderMap = [NSMutableDictionary dictionary];
    });
    
    return instance;
}
@end
#pragma mark - Binder
@implementation Binder
- (void)dealloc
{
    NSLog(@"%@ -- dealloc", NSStringFromClass(self.class));

    [self unBind];
}

+ (instancetype)binderWithModule:(NSString *)module
{
    Binder *b = [[BinderHelper shareBinderHelp].binderMap objectForKey:module];
    if (!b) {
        b = [Binder new];
        b.module = module;
        [[BinderHelper shareBinderHelp].binderMap setObject:b forKey:module];
    }
    return b;
}
+ (void)unBinderWithModule:(NSString *)module
{
    Binder *b = [[BinderHelper shareBinderHelp].binderMap objectForKey:module];
    if (b) {
        [b unBind];
    }
}
- (void)unBind
{
    [self.branches removeAllObjects];
    [self.contexts removeAllObjects];
    [self removeSubBinders];
    
    [[BinderHelper shareBinderHelp].binderMap removeObjectForKey:self.module];
}
- (void)addSubBinder:(Binder *)binder
{
    [self.subBinders addObject:binder];
    binder.superBinder = self;
}
- (void)removeSubBinders
{
    for (id<Binder> binder in self.subBinders) {
        [binder removeSubBinders];
    }
    [self.subBinders removeAllObjects];
}
- (void)removeFromSuperBinder
{
    if ([self.superBinder.subBinders containsObject:self]) {
        [self.superBinder.subBinders removeObject:self];
        self.superBinder = nil;
    }
}
#pragma mark - branch
// bind
- (void)bind:(id<Branch>)branch protocol:(Protocol *)protocol{
    [self bind:branch protocol:protocol identity:nil];
}
- (void)bind:(id<Branch>)branch protocol:(Protocol *)protocol identity:(NSString *)identity
{
    [self.branches setObject:branch forKey:[self getBranchKey:protocol identity:identity]];
}
// get
- (id<Branch>)branch:(Protocol *)protocol {
    return [self branch:protocol identity:nil];
}
- (id<Branch>)branch:(Protocol *)protocol identity:(NSString *)identity
{
    return [self.branches objectForKey:[self getBranchKey:protocol identity:identity]];
}
// unbind
- (void)unbind:(Protocol *)protocol {
    [self unbind:protocol identity:nil];
}
- (void)unbind:(Protocol *)protocol identity:(NSString *)identity
{
    [self.branches removeObjectForKey:[self getBranchKey:protocol identity:identity]];
}

- (NSString *)getBranchKey:(Protocol *)protocol identity:(NSString *)identity
{
    return [NSString stringWithFormat:@"%@%@", NSStringFromProtocol(protocol), identity];
}
- (NSMapTable *)branches
{
    if (!_branches) {
        _branches = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsStrongMemory valueOptions:NSPointerFunctionsWeakMemory];
    }
    return _branches;
}
#pragma mark - context
// add
- (void)addCtx:(id<Context>)ctx
{
    [self addCtx:ctx identity:nil];
}
- (void)addCtx:(id<Context>)ctx identity:(NSString *)identity
{
    [self.contexts setObject:ctx forKey:[self getContextKey:ctx.class identity:identity]];
}
// remove
- (void)removeCtx:(Class)cls
{
    [self removeCtx:cls identity:nil];
}
- (void)removeCtx:(Class)cls identity:(NSString *)identity
{
    [self.contexts removeObjectForKey:[self getContextKey:cls identity:identity]];
}
// get
- (id<Context>)ctx:(Class)cls
{
    return [self ctx:cls identity:nil];
}
- (id<Context>)ctx:(Class)cls identity:(NSString *)identity
{
    id<Context> ctx = [self.contexts objectForKey:[self getContextKey:cls identity:identity]];
    return ctx;
}
- (NSString *)getContextKey:(Class)cls identity:(NSString *)identity
{
    return [NSString stringWithFormat:@"%@%@", NSStringFromClass(cls), identity];
}

@synthesize branches = _branches;
@synthesize contexts = _contexts;
@synthesize module = _module;
@synthesize superBinder = _superBinder;

@synthesize subBinders = _subBinders;
- (NSMutableArray<Binder *> *)subBinders
{
    if (!_subBinders) {
        _subBinders = [NSMutableArray array];
    }
    return _subBinders;
}
- (NSMutableDictionary *)contexts
{
    if (!_contexts) {
        _contexts = [NSMutableDictionary dictionary];
    }
    return _contexts;
}
@end
#pragma mark - BranchRequest
@implementation BranchRequest

@synthesize module;

@synthesize branch;

@synthesize action;

@synthesize target;

@synthesize info;

@end

File: EasyApp/EasyApp/Template/MVC2_swift/MVC2Base.swift
//
//  MVC2Base.swift
//  EasyApp
//
//  Created by 靳小飞 on 2018/5/2.
//  Copyright © 2018年 靳小飞. All rights reserved.
//

import Foundation

private class ModulesManager: NSObject {
    var binderMap: [String: Binder] = [String: Binder]()
    static let shareManager = ModulesManager()
}

class BinderImp: NSObject, Binder{
    var module: String = String()
    var branches: NSMapTable<AnyObject, Branch> = NSMapTable<AnyObject, Branch>()
    var contexts: [String : Context] = [String : Context]()
    var superBinder: Binder?
    var subBinders: [Binder] = [Binder]()
    
    deinit {
        self.unBind()
    }
    
    override init() {
        super.init()
    }
    
    static func binderWithModule(module: String) -> Binder {
        var mod = ModulesManager.shareManager.binderMap[module]
        if mod == nil {
            mod = BinderImp()
            mod?.module = module
            ModulesManager.shareManager.binderMap[module] = mod
        }
        return mod!
    }
    
    static func unBinderWithModule(module: String) {
        let b = ModulesManager.shareManager.binderMap[module]
        b?.unBind()
    }
    
    func unBind() {
        self.branches.removeAllObjects()
        self.contexts.removeAll()
        self.removeSubBinders()
        ModulesManager.shareManager.binderMap.removeValue(forKey: self.module)
    }
    
    func addSubBinder(binder: Binder) {
        self.subBinders.append(binder)
    }
    
    func removeSubBinders() {
        for binder in self.subBinders {
            binder.removeSubBinders()
        }
        self.subBinders.removeAll()
    }
    
    func removeFromSuperBinder() {
        
        if let index = self.superBinder?.subBinders.index(where: { (binder) -> Bool in
            binder.module == self.module
        }) {
            self.superBinder?.subBinders.remove(at: index)
        }
    }
    //
    func bind(branch: Branch, ptl: Protocol) {
        self.bind(branch: branch, ptl: ptl, identity: nil)
    }
    
    func bind(branch: Branch, ptl: Protocol, identity: String?) {
        self.branches.setObject(branch, forKey: self.getBranchKey(ptl, identify: identity) as AnyObject)
    }
    
    func unbind(ptl: Protocol) {
        self.unbind(ptl: ptl, identity: nil)
    }
    
    func unbind(ptl: Protocol, identity: String?) {
        self.branches.removeObject(forKey: self.getBranchKey(ptl, identify: identity) as AnyObject)
    }
    
    func branch(ptl: Protocol) -> Branch? {
        return self.branch(ptl: ptl, identity: nil)
    }
    
    func branch(ptl: Protocol, identity: String?) -> Branch? {
        return self.branches.object(forKey: self.getBranchKey(ptl, identify: identity) as AnyObject)
    }
    //ctx
    func addCtx(context: Context) {
        self.addCtx(context: context, identity: nil)
        
    }
    
    func addCtx(context: Context, identity: String?) {
        self.contexts[self.getContextKey(type(of: context), identify: identity)] = context
    }
    
    func removeCtx(cls: AnyClass) {
        self.removeCtx(cls: cls, identity: nil)
    }
    
    func removeCtx(cls: AnyClass, identity: String?) {
        self.contexts.removeValue(forKey: self.getContextKey(cls, identify: identity))
    }
    
    func ctx(_ cls: AnyClass) -> Context? {
        return self.ctx(cls, identity: nil)
    }
    
    func ctx(_ cls: AnyClass, identity: String?) -> Context? {
        return self.contexts[self.getContextKey(cls, identify: identity)]
    }
    //private
    fileprivate func getBranchKey(_ ptl: Protocol, identify: String?) -> String {
        return String.init(format: "%@%@", NSStringFromProtocol(ptl), identify ?? "")
    }
    fileprivate func getContextKey(_ cls: AnyClass, identify: String?) -> String {
        return String.init(format: "%@%@", NSStringFromClass(cls), identify ?? "")
    }
}

class BranchRequestModel: NSObject, BranchRequest {
    var module: String = String()
    
    var branch: String = String()
    
    var action: Selector?
    
    var target: AnyObject?
    
    var info: AnyObject?
}
